### DISCUSSION TITLE
Running pyhf on Macbook GPUs

### DISCUSSION THREAD

**Discussion by ggustavi** [scikit-hep/pyhf#1489] [Q&A]
Dear pyhf experts and developers,

I was trying to run pyHF (v0.6.1 since I need to set a lower tolerance when running CLs) with jax on GPUs. My first attempt would be trying to run it on my Mac laptop which, as you know, does not support Nvidia GPUs (they are Intel HD Graphics 530 in my case ).

I have figured out that the problem is not related to pyhf itself but actually it’s due to tensorflow with which I cannot see my GPUs.

I tried to install tensorflow based on the guide (https://www.tensorflow.org/install/source) but I did not figure out to install it on Mac actually...is there anyone of you that managed to do it?

I have tried installing https://github.com/apple/tensorflow_macos and with it I am able to run tensorflow but I get conflicts when I try to install pyhf:

```
> pip install --ignore-installed pyhf==0.6.1
…
ERROR: pip’s dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.2.2 requires numpy<1.19.0,>=1.16.0, but you have numpy 1.20.3 which is incompatible.
tensorflow-macos 0.1a3 requires flatbuffers~=1.12.0, but you have flatbuffers 2.0 which is incompatible.
tensorflow-macos 0.1a3 requires grpcio~=1.32.0, but you have grpcio 1.33.2 which is incompatible.
tensorflow-macos 0.1a3 requires numpy~=1.19.2, but you have numpy 1.20.3 which is incompatible.
tensorflow-macos 0.1a3 requires protobuf~=3.13.0, but you have protobuf 3.17.3 which is incompatible.
tensorflow-macos 0.1a3 requires six~=1.15.0, but you have six 1.16.0 which is incompatible.
tensorflow-macos 0.1a3 requires tensorboard~=2.3, but you have tensorboard 2.2.2 which is incompatible.
tensorflow-macos 0.1a3 requires tensorflow-estimator~=2.3.0, but you have tensorflow-estimator 2.2.0 which is incompatible.
tensorflow-macos 0.1a3 requires typing-extensions~=3.7.4, but you have typing-extensions 3.10.0.0 which is incompatible.
```

Did you have to deal with similar issues with your tensorflow setup on MacOS GPUs?
Do you have any suggestion?

Cheers,
Giuliano

**Comment by matthewfeickert** [comment]
Thanks for this very good question @ggustavi. 

Unless I'm mistaken, all the pyhf backends with the exception of PyTorch (but the [AMD GPU support there is in beta](https://pytorch.org/blog/pytorch-for-amd-rocm-platform-now-available-as-python-package/)) require NVIDIA GPUs for native support. I see that you've linked Apple's (now archived) project that advertises GPU support on macOS, but that's for a _currently_ unsupported TensorFlow release for pyhf (given breaking changes that TensorFlow made we have to figure out how to reconcile). c.f. Issue #293. So I think if you want to use your Mac's GPUs you're out of luck. :disappointed: 

In the future you might be interested to know that we're also distributing CUDA enabled Docker images: https://github.com/pyhf/cuda-images

e.g. 

```
docker pull pyhf/cuda:0.6.1-jax-cuda-11.1
```

**Reply by ggustavi** [reply]
Hi @matthewfeickert,
thanks for the reply. Sorry if my question is naive but I don't completely get how the mentioned docker image can help me to use the MAC GPUs.
Thank you.
Cheers,
Giuliano

**Reply by matthewfeickert** [reply]
> Sorry if my question is naive but I don't completely get how the mentioned docker image can help me to use the MAC GPUs.

Oh, no I just wasn't clear enough in what I mean. This doesn't help your local MacOS problem at all, but in the situation where you hopefully also have some access to GPUs on a cluster or HPC environment these images provide an out-of-the-box environment.
